### DISCUSSION TITLE
Is there an API to monitor changes in all Router and Child nodes in a Thread network?

### DISCUSSION THREAD

**Discussion by xiaojiang2017** [openthread/openthread#10572] [Q&A]
Hi Experts, 
We are doing some work on Thread network diagnostics. Is there an API to monitor changes in all Router and Child nodes in a Thread network? How do I know if the children of other router nodes have changed, such as leaving the network or reattaching to the network?
Thanks.

**Comment by jwhui** [comment]
Thread provides a way to query topology state from other devices. You can use this to periodically query state over time and track changes.

You may find the [`meshdiag`](https://github.com/openthread/openthread/blob/main/src/cli/README.md#meshdiag-topology-ip6-addrs-children) CLI command useful.

**Reply by xiaojiang2017** [reply]
Thanks @jwhui . I'm concerned that frequent calls to `meshdiag` might affect network performance. So I think that if TBR can know that there are node change events in the network, such as the addition and removal of Routers, it will actively call `meshdiag` once.

**Reply by abtink** [reply]
Some info which may help you:

- Check the existing [`Notifier` events](https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/instance.h#L201-L202): `OT_CHANGED_THREAD_CHILD_ADDED` and `OT_CHANGED_THREAD_CHILD_REMOVED`
- There is also `otThreadRegisterNeighborTableCallback()`
https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/thread_ftd.h#L789-L839
- Note that the above mechanisms apply only to the **device's immediate neighbors** (those with a direct link).
-  _There's currently no mechanism/API to be notified of changes to remote devices, i.e., children of other routers, or routers that are multiple hops away and not direct neighbors._ "MeshDiag" provides mechanism to query and learn.

------------

OpenThread also provides "HistoryTracker" feature (can be enabled on more capable devices with more resources):
https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/history_tracker.h#L39-L44

It can track changes to "route table" which tells us when/if any router (neighbor or remote) is added/removed:

https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/history_tracker.h#L195-L222

https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/history_tracker.h#L373-L390

It also tracks history of changed to neighbors (child/router):
https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/history_tracker.h#L163-L194

**Reply by xiaojiang2017** [reply]
Cool, that's what I wanted. Thanks.

**Reply by jwhui** [reply]
Note that Thread intentionally does not require any device to maintain full topology information for the entire Thread network, for scalability reasons. So obtaining any state from another node in the network will always incur additional Thread traffic.
